pmedian.jl

```diff
@@ -0,0 +1,269 @@
+import MathOptInterface
+import SCS
+import GLPK
+import Random
+
+using SparseArrays
+using TimerOutputs
+
+const MOI = MathOptInterface
+
+struct PMedianData
+    num_facilities::Int
+    num_customers::Int
+    num_locations::Int
+    customer_locations::Vector{Float64}
+end
+
+# This is the LP relaxation.
+function generate_moi_problem(model, data::PMedianData)
+    facility_variables = MOI.add_variables(model, data.num_locations)
+    for v in facility_variables
+        MOI.add_constraint(model, v, MOI.GreaterThan(0.0))
+        MOI.add_constraint(model, v, MOI.LessThan(1.0))
+    end
+    assignment_variables = [MOI.add_variable(model) for i in 1:data.num_customers, j in 1:data.num_locations]
+    for v in assignment_variables
+        MOI.add_constraint(model, v, MOI.GreaterThan(0.0))
+        # "Less than 1.0" constraint is redundant.
+    end
+    objective = MOI.ScalarAffineFunction(
+        [MOI.ScalarAffineTerm(abs(data.customer_locations[i] - j), assignment_variables[i,j])
+        for i in 1:data.num_customers for j in 1:data.num_locations],
+            0.0)
+    MOI.set(model, MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}(), objective)
+    MOI.set(model, MOI.ObjectiveSense(), MOI.MIN_SENSE)
+    for i in 1:data.num_customers, j in 1:data.num_locations
+        # assignment_variables[i,j] <= facility_variables[j]
+        MOI.add_constraint(model, MOI.ScalarAffineFunction(
+            [MOI.ScalarAffineTerm(1.0, assignment_variables[i,j]),
+            MOI.ScalarAffineTerm(-1.0, facility_variables[j])], 0.0),
+            MOI.LessThan(0.0))
+    end
+    for i in 1:data.num_customers
+        # sum_j assignment_variables[i,j] = 1
+        MOI.add_constraint(model, MOI.ScalarAffineFunction(
+            [MOI.ScalarAffineTerm(1.0, assignment_variables[i,j]) for j in 1:data.num_locations],
+            0.0),
+            MOI.EqualTo(1.0)
+        )
+    end
+    # sum_j facility_variables[j] = num_facilities
+    MOI.add_constraint(model, MOI.ScalarAffineFunction(
+        MOI.ScalarAffineTerm.(1.0, facility_variables), 0.0),
+        MOI.EqualTo{Float64}(data.num_facilities)
+    )
+    return assignment_variables, facility_variables
+end
+
+function generate_glpk_problem(prob, data::PMedianData)
+    facility_variables = [GLPK.glp_add_cols(prob, 1) for i in 1:data.num_locations]
+    assignment_variables = [GLPK.glp_add_cols(prob, 1) for i in 1:data.num_customers, j in 1:data.num_locations]
+    for j in 1:data.num_locations
+        GLPK.glp_set_col_bnds(prob, facility_variables[j], GLPK.GLP_DB, 0.0, 1.0)
+    end
+    for i in 1:data.num_customers, j in 1:data.num_locations
+        GLPK.glp_set_col_bnds(prob, assignment_variables[i, j], GLPK.GLP_DB, 0.0, 1.0)
+        GLPK.glp_set_obj_coef(prob, assignment_variables[i, j], abs(data.customer_locations[i] - j))
+    end
+    GLPK.glp_set_obj_dir(prob, GLPK.GLP_MIN)
+
+    I = Cint[0] # Extra padding because GLPK starts reading from the second element (??)
```

Comment:
Yes. It's so they can have 1-based indexing in C. Guess it simplifies the implementation???

You can also use `offset(x)` in the actual C call to GLPK.

See: https://github.com/jump-dev/GLPK.jl/blob/41e649cd81864e90065ac78f42317edcebcd9951/src/GLPK.jl#L20-L32

Reply:
I left this as is for simplicity, to avoid depending on internal helpers in GLPK.jl.